subsys/testsuite/ztest: create stub headers at CMake configure time

ycsin opened the pull request:
These header stubs seems to be required by the `unit_testing` board only. Instead of having them in the source, let's remove them and only generate them at CMake configure time.

~~See also #69442, which relocates these headers instead of having them generated at build~~

ycsin committed:
subsys/testsuite/ztest: create stub headers at CMake configure time

Create unittest stub headers at CMake configure time.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented on subsys/testsuite/ztest/include/zephyr/ztest.h:
```diff
@@ -22,19 +22,6 @@
 #error "You need to add CONFIG_ZTEST to your config file."
 #endif
 
-#ifndef KERNEL
-#define ARCH_STACK_PTR_ALIGN 8
-/* FIXME: Properly integrate with Zephyr's arch specific code */
-#ifdef __cplusplus
-extern "C" {
-#endif
-struct esf;
-typedef struct esf z_arch_esf_t;
-#ifdef __cplusplus
-}
-#endif
-#endif /* KERNEL */
-
```
removed as they don't seem required by upstream CI, can be split into another commit

ycsin reviewed (commented):
ping

tejlmand commented on cmake/modules/unittest.cmake:
```diff
@@ -55,6 +55,15 @@ include(${ZEPHYR_BASE}/cmake/kobj.cmake)
 add_dependencies(test_interface ${KOBJ_TYPES_H_TARGET})
 gen_kobj(KOBJ_GEN_DIR)
 
+FILE(TOUCH ${APPLICATION_BINARY_DIR}/zephyr/include/generated/offsets.h)
```
Minor comment related to commit message.

This is not done at build time, but at CMake configure time.

ycsin replied:
Fixed, thanks

tejlmand commented on cmake/modules/unittest.cmake:
```diff
@@ -55,6 +55,15 @@ include(${ZEPHYR_BASE}/cmake/kobj.cmake)
 
+FILE(TOUCH ${APPLICATION_BINARY_DIR}/zephyr/include/generated/offsets.h)
+FILE(TOUCH ${APPLICATION_BINARY_DIR}/zephyr/include/generated/syscall_list.h)
+FILE(TOUCH ${APPLICATION_BINARY_DIR}/zephyr/include/generated/devicetree_generated.h)
```
other CMake function calls are written in lower case, so please adjust to `file(...)` if you continue this approach.

ycsin replied:
How did I miss that? Thanks

tejlmand commented on include/zephyr/linker/linker-defs.h:
```diff
@@ -25,14 +25,7 @@
- * DT_NODE_HAS_STATUS.
- */
-#ifdef ZTEST_UNITTEST
-#define DT_NODE_HAS_STATUS(node, status) 0
```
Curious, where is `DT_NODE_HAS_STATUS(node, status)` defined now / why is it no longer needed ?

ycsin replied:
This change is probably not required, removed.

ycsin committed:
ztest: include: unittest: relocate cpu.h and include only if unit-test

The `cpu.h` is meant to be a hack around Zephyr's dependencies
for unit-test, relocate it to a folder that's included only by
unit-test.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented on cmake/modules/unittest.cmake:
```diff
@@ -58,8 +58,22 @@ include(${ZEPHYR_BASE}/cmake/kobj.cmake)
 gen_kobj(KOBJ_GEN_DIR)
 
+# Generates empty header files to build
+file(TOUCH ${APPLICATION_BINARY_DIR}/zephyr/include/generated/devicetree_generated.h)
```
is it more elegant to do a list + foreach instead?

nashif replied:
yes, I guess.

ycsin replied:
converted to list + foreach

ycsin commented:
ping - @nashif could you PTAL again? Thanks!

ycsin commented:
another ping @nashif
